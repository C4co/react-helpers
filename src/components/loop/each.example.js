import React from "react"
import styled from "styled-components"
import { Each } from "./loop.item"

const Wrapper = styled.section`
  width: 100%;
  max-width: 500px;
  margin: 0 auto;
`

function Example(){
  return(
    <Wrapper>
      <Each items={["first", "second", "third"]}>
        {(item, index) => (
          <section>
            <h1> {index} -> {item} </h1>
            <hr/>
          </section>
        )}
      </Each>
    </Wrapper>
  )
}

export default Example