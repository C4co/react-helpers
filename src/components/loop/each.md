## Each

Iteration over multiple itens.

#### Usage
```jsx
  import { Each } from "react-utl"
```

#### Example:

```jsx
  <Each items={["first", "second", "third"]}>
    {(item, index) => (
      <section>
        <h1> {index} -> {item} </h1>
        <hr/>
      </section>
    )}
  </Each>
```

#### Properties:

| Name | Type | Default |
| --- | --- | --- |
| @items | array | - |
| @children | function(item, index) | - |
