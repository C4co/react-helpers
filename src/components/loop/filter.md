## Filter

Iteration over multiple itens and filter items based in a condition.

#### Usage
```jsx
  import { Filter } from "react-utl"
```

#### Example:

```jsx
  <Filter items={[1, 2, 3]} if={item => item > 1}>
    {(item, index) => (
      <section>
        <h1> Filtered value => {index} - {item} </h1>
      </section>
    )}
  </Filter>
```

#### Properties:

| Name | Type | Default |
| --- | --- | --- |
| @items | array | [] |
| @children | function(item, index) | - |
| @if | function(item) | - |
