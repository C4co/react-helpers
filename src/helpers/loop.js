import React from "react"

function ID() {
  return '_' + Math.random().toString(36).substr(2, 9);
}

/*
  name: Each
  @items -> array - list of elements
  @children -> function - callback passing all elements of array
*/
const Each = props => {
  return (
    <>
      {props.items.map((item, index, arr) => {
        return (
          <React.Fragment key={ID()}>
            {props.children(item, index, arr)}
          </React.Fragment>
        )
      })}
    </>
  )
}

/*
  name: Filter
  @items -> array - list of elements
  @children -> function - callback passing all elements filtered by condition.
*/
const Filter = props => {
  const filtered = props.items.filter((item, index, arr) =>
    props.if(item, index, arr)
  )

  return (
    <>
      {filtered.map((item, index, arr) => {
        return (
          <React.Fragment key={ID()}>
            {props.children(item, index, arr)}
          </React.Fragment>
        )
      })}
    </>
  )
}

export { Each, Filter }
