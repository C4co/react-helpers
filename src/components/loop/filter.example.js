import React from "react"
import styled from "styled-components"
import { Filter } from "./loop.item"

const Wrapper = styled.section`
  width: 100%;
  max-width: 500px;
  margin: 0 auto;
`

function Example(){
  return(
    <Wrapper>
      <Filter items={[1, 2, 3]} if={item => item > 1}>
        {(item, index) => (
          <section>
            <h1> Filtered value => {index} - {item} </h1>
          </section>
        )}
      </Filter>
    </Wrapper>
  )
}

export default Example