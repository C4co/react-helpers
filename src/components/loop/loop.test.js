import { mount } from "enzyme"
import { Each, Filter } from "./loop.item"
import React from "react"

describe("Loop component", () => {
  test("<Each>: should iterate over each array item.", () => {

    const each = mount(
      <Each items={['first', 'second', 'third']}>
        {(item, index) => (
          <h1>
            {index}->{item}
          </h1>
        )}
      </Each>
    )

    expect(each.contains(<h1>0->first</h1>)).toBe(true)
    expect(each.contains(<h1>1->second</h1>)).toBe(true)
    expect(each.contains(<h1>2->third</h1>)).toBe(true)
    expect(each).toMatchSnapshot()
  })

  test("<Filter>: Should iterate and filter an array based in a condition", () => {
    const filter = mount(
      <Filter items={[1, 2, 3]} if={item => item > 1}>
        {(item, index) => (
          <h1>
            {index}->{item}
          </h1>
        )}
      </Filter>
    )

    expect(filter.contains(<h1>0->2</h1>)).toBe(true)
    expect(filter.contains(<h1>1->3</h1>)).toBe(true)
  })
})
