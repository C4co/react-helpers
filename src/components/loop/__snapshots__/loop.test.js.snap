// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loop component <Each>: should iterate over each array item. 1`] = `ReactWrapper {}`;
